### DISCUSSION TITLE
Bidir compatibility.

### DISCUSSION THREAD

**Discussion by teegee** [zeroc-ice/ice#2495] [Q&A] [answered]
We're looking into using bidir connections for a project because of firewall issues for callback proxies. This works fine, following your [docs](https://doc.zeroc.com/ice/3.7/client-server-features/connection-management/bidirectional-connections), but the issue is that we will encounter older clients that are not aware that the server wants to use bidir connections. As per the documentation, this doesn't work:

> If the client forgot to call setAdapter on the connection, the notify call on the fixed proxy fails with an ObjectNotExistException raised by the client and propagated to the server.

..which is exactly what happens when an older client connects. My question: is there is a way that server-side could detect whether the client has sent us a callback proxy that is bidir-enabled? That way, for older clients the server could just not setup for bidir proxies, and things would work as before (if the firewall permits).

Cheers!

    Rob

**Comment by pepone** [comment, marked as answer]
Hi Rob,

You can check the proxy endpoints to handle the situation where older clients are unaware of bidirectional connections. Older clients will send proxies containing endpoints, while proxies sent by new clients will not contain endpoints.

You can use the `ice_getEndpoints` proxy method to determine if the proxy contains endpoints.

Cheers,
Jose

**Reply by teegee** [reply]
Hi Jose,

Thanks for the tip.. I tried that, but the proxy still has endpoints for new clients, since I didn't change the adapter creation process (client still uses a configured adapter, even though the endpoints won't be used). Do I have to use an 'empty' adapter like in the docs, or perhaps use ice_endpoints() to strip them from the callback proxy before passing it to the server?

Cheers!

Rob

**Reply by pepone** [reply]
I would use an adapter without endpoints like in the docs. But striping the endpoints should also work.

**Reply by teegee** [reply]
Thanks Jose :)
